Library/ViewModels/PledgePaymentMethodsViewModelTests.swift

```diff
@@ -1786,4 +1786,231 @@ final class PledgePaymentMethodsViewModelTests: TestCase {
+    let cards = UserCreditCards.withCards([])
+
+    let graphUser = GraphUser.template |> \.storedCards .~ cards
+    let graphUserResponse = UserEnvelope<GraphUser>(me: graphUser)
+    let addPaymentSheetResponseVisa = CreatePaymentSourceEnvelope(
+      createPaymentSource: .init(isSuccessful: true, paymentSource: UserCreditCards.visa)
+    )
+    let addPaymentSheetResponseDiscover = CreatePaymentSourceEnvelope(
+      createPaymentSource: .init(isSuccessful: true, paymentSource: UserCreditCards.discover)
+    )
+
+    let mockService1 = MockService(
+      addPaymentSheetPaymentSourceResult: .success(addPaymentSheetResponseVisa),
+      fetchGraphUserResult: .success(graphUserResponse)
+    )
+
+    let mockService2 = MockService(
+      addPaymentSheetPaymentSourceResult: .success(addPaymentSheetResponseDiscover)
+    )
+
+    let project = Project.template
+      |> \.availableCardTypes .~ ["VISA"]
+
+    withEnvironment(apiService: mockService1, currentUser: User.template) {
+      self.vm.inputs
+        .configure(with: (
+          User.template,
+          project,
+          "checkoutID",
+          Reward.template,
+          .pledge,
+          .discovery
+        ))
+      self.vm.inputs.viewDidLoad()
+
+      self.scheduler.run()
+
+      // There are no cards
+      XCTAssertEqual(self.reloadPaymentMethodsIsSelected.lastValue, [])
+      XCTAssertEqual(self.reloadPaymentMethodsAvailableCardTypes.lastValue, [])
+      XCTAssertNil(self.reloadPaymentMethodsSelectedCardId.lastValue ?? nil)
+
+      // Add a new card (the visa)
+      self.vm.inputs
+        .paymentSheetDidAdd(
+          clientSecret: "si_fake",
+          paymentMethod: "pm_fake"
+        )
+
+      self.scheduler.run()
+
+      // There is now one card, in the new section
+      XCTAssertEqual(self.reloadPaymentSheetPaymentMethodsCards.lastValue?.count, 1)
+      XCTAssertEqual(self.reloadPaymentMethodsCards.lastValue?.count, 0)
+
+      // The newly added card is enabled and selected
+      let newCardData = self.reloadPaymentSheetPaymentMethodsCards.lastValue?.first
+      XCTAssertEqual(newCardData?.isEnabled, true)
+      XCTAssertEqual(newCardData?.isSelected, true)
+      XCTAssertEqual(self.reloadPaymentMethodsSelectedCardId.lastValue, UserCreditCards.visa.id)
+    }
+
+    withEnvironment(apiService: mockService2, currentUser: User.template) {
+      // Add another new card (the discover)
+      self.vm.inputs
+        .paymentSheetDidAdd(
+          clientSecret: "si_fake",
+          paymentMethod: "pm_fake"
+        )
+
+      self.scheduler.run()
+
+      // There are now twos card, in the new section
+      XCTAssertEqual(self.reloadPaymentSheetPaymentMethodsCards.lastValue?.count, 2)
+      XCTAssertEqual(self.reloadPaymentMethodsCards.lastValue?.count, 0)
+
+      // The discover is disabled and not selected.
+      let discoverCardData = self.reloadPaymentSheetPaymentMethodsCards.lastValue?.first(where: { (
```

Comment:
Can someone remind me if there's a more beautiful way to do this search? Maybe with a keypath?

Reply:
FWIW I did this because the cards weren't ordered the way I expected, but I also realized it doesn't _matter_ what order two disabled cards are in.

Reply:
I don't know a nicer way 😭 Tbh, if it's deterministic, I think I'd just go by the index (which would mean that the test would need to be changed if we change the order in the future, so I think I do like this search better).